packages/grafana-runtime/src/analytics/plugins/usePluginInteractionReporter.ts

```diff
@@ -0,0 +1,38 @@
+import {
+  type DataSourcePluginContextType,
+  isDataSourcePluginContext,
+  type PluginContextType,
+  usePluginContext,
+} from '@grafana/data';
+
+import { reportInteraction } from '../utils';
+
+import { createDataSourcePluginEventProperties, createPluginEventProperties } from './eventProperties';
+
+export function usePluginInteractionReporter(): typeof reportInteraction {
+  const context = usePluginContext();
+
+  if (isDataSourcePluginContext(context)) {
+    return createDataSourceReporter(context);
+  }
+
+  return createPluginReporter(context);
+}
+
+function createDataSourceReporter(context: DataSourcePluginContextType): typeof reportInteraction {
+  const { meta, dataSource } = context;
+  const info = createDataSourcePluginEventProperties(meta, dataSource);
+
+  return (interactionName: string, properties?: Record<string, unknown>) => {
+    return reportInteraction(interactionName, { ...properties, ...info });
+  };
+}
+
+function createPluginReporter(context: PluginContextType): typeof reportInteraction {
+  const { meta } = context;
+  const info = createPluginEventProperties(meta);
+
+  return (interactionName: string, properties?: Record<string, unknown>) => {
+    return reportInteraction(interactionName, { ...properties, ...info });
+  };
+}
```

Comment:
I actually like how the code is laid out currently, so this is just a quick idea on how we could make it a bit more DRY, I will leave it up to you. (I am not sure if TS would complain here, though)

```javascript
export function usePluginInteractionReporter(): typeof reportInteraction {
  const context = usePluginContext();
  
  return createPluginReporter(context);
}

function createPluginReporter(context: PluginContextType): typeof reportInteraction {
  const { meta } = context;
  const info = isDataSourcePluginContext(context) ?
    createDataSourcePluginEventProperties(meta, context.dataSource) :
    createPluginEventProperties(meta);
    
  return (interactionName: string, properties?: Record<string, unknown>) => {
    return reportInteraction(interactionName, { ...info, ...properties });
  };
}
```

Reply:
not sure I like doing all this for every render/call of usePluginInteractionReporter, would not this be cleaner? 

```tsx
function MyPluginLink() {
   const context = usePluginContext();

   const doStuff = () => {
      context.reportInteraction("Button clicked", {asd: true}); 
   }
 
  return <Button onClick={doStuff}>Hello</Button>
}
```

Reply:
the smaller surface area (number of functions, hooks, types) the plugin needs to depend on the better

Reply:
hmm, I don't really agree. Isn't it better to have smaller specific hooks that you can consume instead of having one bigger that contains a lot more?

I'm thinking that we probably will add more things (potentially state) to the context in the future and probably don't want plugin developers to consume that directly but instead via a hook.